pangeo_forge_recipes/transforms.py

```diff
@@ -443,37 +435,89 @@ class CombineReferences(beam.PTransform):
+            remote_protocol=self.remote_protocol,
+            **self.mzz_kwargs,
+        )
+
+    def handle_gribs(self, indexed_references: Tuple[Index, list[dict]]) -> Tuple[Index, dict]:
+        """Handles the special case of GRIB format files by combining multiple references."""
+
+        references = indexed_references[1]
+        idx = indexed_references[0]
+        if len(references) > 1:
+            ref = self.to_mzz(references).translate()
+            return (idx, ref)
+        elif len(references) == 1:
+            return (idx, references[0])
+        else:
+            raise ValueError("No references produced for {idx}. Expected at least 1.")
+
+    def bucket_by_position(
+        self, indexed_references: Tuple[Index, dict], global_position_min_max_count
+    ) -> Tuple[int, dict]:
+        """Assigns a bucket based on index position to help preserve data order."""
+        idx = indexed_references[0]
+        ref = indexed_references[1]
+        global_min, global_max, global_count = global_position_min_max_count
+        sort_dimension = self.concat_dims[-1]
+        position = idx.find_position(sort_dimension)
+        range_size = global_max - global_min
+        num_buckets = math.ceil(global_count / self.max_refs_per_merge)
+        bucket_size = range_size / num_buckets
+        bucket = int((position - global_min) / bucket_size)
+        return bucket, ref
+
+    def global_combine_refs(self, refs) -> fsspec.FSMap:
+        """Performs a global combination of references to produce the final dataset."""
+        return fsspec.filesystem(
+            "reference",
+            fo=self.to_mzz(refs).translate(),
+            storage_options={
+                "remote_protocol": self.remote_protocol,
+                "skip_instance_cache": True,
+            },
+        ).get_mapper()
+
+    def expand(self, reference_lists: beam.PCollection) -> beam.PCollection:
+        min_max_count_positions = (
+            reference_lists
+            | "Get just the positions"
+            >> beam.MapTuple(lambda k, v: k.find_position(self.concat_dims[-1]))
+            | "Get minimum/maximum positions" >> beam.CombineGlobally(MinMaxCountCombineFn())
+        )
+        return (
+            reference_lists
+            | "Handle special case of gribs" >> beam.Map(self.handle_gribs)
+            | "Bucket to preserve order"
+            >> beam.Map(
+                self.bucket_by_position,
+                global_position_min_max_count=beam.pvalue.AsSingleton(min_max_count_positions),
+            )
+            | "Group by buckets for ordering" >> beam.GroupByKey()
+            | "Distributed reduce" >> beam.MapTuple(lambda k, refs: self.to_mzz(refs).translate())
+            | "Assign global key for collecting to executor" >> beam.Map(lambda ref: (None, ref))
+            | "Group globally" >> beam.GroupByKey()
+            | "Global reduce" >> beam.MapTuple(lambda k, refs: self.global_combine_refs(refs))
```

Comment:
So just for my own understanding, what I think is happening is this:
1. In the case of grib, merge multiple references so we start from a flat list of references
2. The next map operation buckets the references while preserving their order, such that each bucket will have a set of ordered references, ordered by the last `concat_dim` dimension.
3. The first GroupByKey ensures the order of the buckets is preserved.
4. Create a multizarrtozarr ref for each bucket (`to_mzz`)
5. This part I'm less clear about, from what I can tell we're assigning a new key to each group - but is each group a group of single references or merged references?
6. Merge everything and write to the target.

Reply:
The global reduce here is kind of weird, I agree. I would love a `beam.Collect` or something. So it goes. What we have to do, so far as I can tell, is to assign *the same key* to all records if we want to ensure that they are realized together on one executor

Reply:
> Merge everything and write to the target.

Just want to point out that nothing is being written in `global_combine_refs` just setting up an `fsspec` target for the next transform that does the writing

Reply:
> assign the same key to all records (None here) and then beam.GroupByKey if we want to ensure that they are realized together on one executor

Thanks that is super helpful explanation. In that case how are we assured that the final combine is done in order on the last single executor?

Reply:
Experiments I've run locally w/ `multiprocessing` appear to demonstrate that the list fed to `MultiZarrToZarr` can have any ordering at all so long as there exists some ordering of the elements that would be contiguous. It is when you have separate `mzz` calls that you run into the risk of one of the sub-lists not having contiguity among its would-be merged references
Messing with ordering here should illustrate the behavior: https://gist.github.com/moradology/0e207a7ea4afd7251b0df604ac10958e#file-wtf-py-L69-L71